Hook rkt kubelet runtime up to network plugins

dcbw opened the pull request.

dcbw committed:
Allow using netns path instead of container PID to change hairpin mode

dcbw committed:
kubelet/runtime: add method to return pod network namespace path

Some runtimes (eg, Hypernetes) don't create network namespaces for pods,
so network plugins must correctly handle any error returned from the
runtime.

dcbw committed:
rkt: set up network namespace with plugins

dcbw commented:
@yifan-gu @euank @bprashanth here's the direction I think the whole rkt/CNI/kubenet ball of wax should go...  let me know what you think!

dcbw commented:
`CONTAINER_RUNTIME=rkt RKT_PATH=/usr/bin/rkt RKT_STAGE1_IMAGE=/usr/libexec/rkt/stage1-host.aci LOG_LEVEL=5 NET_PLUGIN=kubenet hack/local-up-cluster.sh -o _output/local/bin/linux/amd64/`

on Fedora 23 with F24's rkt RPM (1.3.0+gitdd7aa64) runs pods, configures the network, and cleans up when the pod terminates for me.  If others could test too, preferably with whatever rkt development version rktnetes will get based on, that would be awesome...

euank commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -925,6 +968,34 @@ func (r *Runtime) preparePod(pod *api.Pod, pullSecrets []api.Secret) (string, *k
+		}
+
+		// Run pod in the namespace we just created and set up
+		runPrepared = fmt.Sprintf("%s netns exec %s %s", r.ipPath, uuid, runPrepared)
```
I found that `ip netns exec` breaks on some systems (e.g. my development laptop, running a fairly modern gentoo install):

```
$ sudo ip netns add foo
$ sudo ip netns exec foo rkt run --net=host docker://busybox --exec=ls -- /sys/fs/cgroup
image: using image from local store for image name coreos.com/rkt/stage1-coreos:1.4.0+gitb7589f9
image: using image from local store for url docker://busybox
stage1: error calling sd_pid_get_owner_uid: exec format error
$ sudo ip netns exec foo ls /sys/fs/cgroup
# nothing
$ ip -V
ip utility, iproute2-ss160111
```

`nsenter` didn't have that issue for me and `nsenter` is already depended on for `ExecInContainer`.

I can't reproduce this issue on CoreOS (`ip -V -> iproute2-ss150210`)

Ref: https://bugzilla.redhat.com/show_bug.cgi?id=882047

That's the reason I used `nsenter` for execing rather than `ip`. I don't know the state of this bug on other distros, but if you're confident it doesn't affect any real users (only the fake users running gentoo :) or if it's a kernel misconfiguration on my end or such, I'm happy to be persuaded. Otherwise, nsenter seems like the safer way to be sure this isn't an issue.

dcbw commented:
@euank hmm, interesting.  For core tools like that, I tend to think that if the tool is broken, that's not really Kubernetes problem and that tool should get fixed, and will get fixed.  Of course if it impacts a wide audience, then we should work around the problem until it's fixed.

dcbw commented:
@euank not opposed to using nsenter, just wondering what versions of iproute2 had that bug, and how widely that version is deployed before falling back to nsenter.

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1320,6 +1393,13 @@ func (r *Runtime) KillPod(pod *api.Pod, runningPod kubecontainer.Pod) error {
 	}
 
+	if !kubecontainer.IsHostNetworkPod(pod) {
+		if err := r.cleanupPodNetwork(&runningPod, true); err != nil {
```
@dcbw This takes care of the network cleanup when we explicitly kill/stop the pods. But what if the pod is a crash loop, and is deleted by garbage collection?
Actually, I don't find any place in dockertools that take care of the network when the pod is garbage collected today, did I miss something?

yifan-gu replied:
@dcbw I guess the docker pod works fine today because the `pause` container hardly dies, so in most case the only time it dies is when it's been explicitly killed, which invokes the `networkplugin.TearDown()` accordingly.

dcbw commented:
@euank looks like the iproute2 patch that fixes the issue is 144e6ce1679a768e987230efb4afa402a5ab58ac and that's been part of 3.8 and later, which was released on 2013-02-21, eg a really long time ago....

dcbw commented:
@yifan-gu I was wondering about the GC code, yeah.  Are there other places that should clean up the netns besides the GC code?

euank commented:
It's possible I'm encountering a separate bug actually since I'm on 4.4.0... I'll look into it a bit more.

dcbw commented:
@euank sure; if you can figure out what's going on there that would be great. But it's looking like we should probably use nsenter as you suggest, at least for now.  I'll make that change tomorrow and repush the PR.

yifan-gu commented:
> @yifan-gu I was wondering about the GC code, yeah. Are there other places that should clean up the netns besides the GC code?

I don't think so. cc @kubernetes/sig-node @yujuhong @Random-Liu @dchen1107 https://github.com/kubernetes/kubernetes/pull/25062#discussion_r61829829

feiskyer commented on pkg/kubelet/container/runtime.go:
```diff
@@ -94,6 +94,8 @@ type Runtime interface {
 	// Returns Image statistics.
 	ImageStats() (*ImageStats, error)
+	// Returns the filesystem path of the pod's network namespace
+	GetNetNS(containerID ContainerID) (string, error)
```
Is it essential for adding `GetNetNS()` to `Runtime` interface? Not all container runtimes (e.g., HyperContainer, which is a hypervisor-based container runtime) running containers inside network namespace.

dcbw replied:
> Is it essential for adding GetNetNS() to Runtime interface? Not all container runtimes (e.g., HyperContainer, which is a hypervisor-based container runtime) run containers inside network namespace.

@feiskyer Good point, though I'd like to reduce the usage of casting in the network plugins and potentially make them usable for more runtimes.  What if we specified in the Runtime interface that a Runtime was not required to return a valid netns path if that runtime does not do namespace creation?  Network plugins would then need to check if the returned path was "" and return an error if they could not continue without a path.

euank replied:
I think it's cleaner to have a `NetworkPluggableRuntime` (name?) interface which just has `GetNetNS` since that's the only method needed, and then docker & rkt can implement it and you can cast to that, rather than a concrete implementation.
If some runtimes will stub out the method, it's not really part of a container runtime.

dcbw replied:
I don't really see the difference right now between GetNetNS() and a NetworkPluggableRuntime.  If we add more methods that runtimes won't implement, then sure, maybe we re-evaluate.

Also note that the only reason there *is* a GetRuntime() is for GetNetNS(); CNI uses it for Status() too but I think that's bogus and it should follow the same model as kubenet does, but that's another PR.  I think we can probably get rid of GetRuntime() in favor of some other interface that the runtime actually implements.  Even better, since the runtimes themselves are calling the functions, we should just pass an interface into them (or nil) instead of having a roundable GetRuntime().

In summary, I'd like to go with GetNetNS() for now, and rework GetRuntime() in a further PR if that's OK?  With the understanding that yes, not all runtimes support returning a netns path now or in the future.

euank replied:
The difference is indeed just conceptual/academic right now, not practical, so I'm fine with this for now with the hope of an even better solution later. Thanks for the explanation as well.

euank commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1424,6 +1504,20 @@ func (r *Runtime) SyncPod(pod *api.Pod, podStatus api.PodStatus, internalPodStat
 }
 
+func netnsPathFromUuid(uuid string) string {
+	return fmt.Sprintf("/var/run/netns/%s", uuid)
```
It might be nice to prepend `k8-rkt-` or something so a user can easily identify namespaces we created. Fine with it as is as well.

euank commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -863,11 +883,34 @@ func (r *Runtime) generateRunCommand(pod *api.Pod, uuid string) (string, error)
+		}
+	}
+
+	if err := exec.Command("ip", "netns", "del", podID.uuid).Run(); err != nil {
```
nit, `r.execer` for mocking would be nice.

euank commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -863,11 +898,34 @@ func (r *Runtime) generateRunCommand(pod *api.Pod, uuid string) (string, error)
+	if teardown {
+		glog.V(3).Infof("Calling network plugin %s to tear down pod for %s", r.networkPlugin.Name(), kubecontainer.BuildPodFullName(pod.Name, pod.Namespace))
+		pluginPodID := buildContainerID(&containerID{podID.uuid, pod.Name})
+		if err := r.networkPlugin.TearDownPod(pod.Namespace, pod.Name, pluginPodID); err != nil {
```
I had this stuff (add & plugin) before `prepare` because at prepare time environment variables are finalized, and one goal I have for this is supporting the downward api (including environment variable of the podIP). The idea to me was if you setup the namespace and run the plugins prior to preparing, you can inject the IP into prepare for `/etc/hosts` file generation and downwards api.

yifan-gu replied:
@euank If you run the plugins before prepare. How did you get the uuid of the rkt? Or are you just using some random uuid?

dcbw replied:
@yifan-gu the Kubernetes pod UID gets used instead of the rkt UUID.  The Kubernetes UID is already seen extensively in kubelet logs and the API, so I think it's OK to use it instead of the rkt UUID.  With the latest commits, the Kubernetes UID is also written to the service file, so it's possible to correlate the two without looking at logs.

dcbw commented:
@euank I think I've addressed all your comments, PTAL thanks!

To ensure we clean up pods when garbage collecting we now write out the [X-Kubernetes] section of the service file, which there were already unused constants for.  That gets parsed by the garbage collection code to get the pod name, namespace, and ID.  Since the network setup is now done before rkt prepare, we can no longer use the rkt UUID but instead must use the Kubernetes pod UID.

dcbw commented:
@feiskyer also updated the code documentation for GetNetNS() to make it clear that runtimes don't have to return a valid value.

feiskyer commented:
> @feiskyer also updated the code documentation for GetNetNS() to make it clear that runtimes don't have to return a valid value.

@dcbw Thanks. Adding `GetNetNS()` indeed works for today's docker and rkt runtime, but the network plugin built by this may be not working for new runtimes. It's better to add netns param together with runtime-type in `networkPlugin.SetUpPod`, so network plugins could process more generally.

euank commented:
@dcbw Thanks! I like adding the UID into the service file. I'm going to do some quick manual testing (local-up & gce+coreos+rkt) and make sure things look sane, and if so lgtm.

dcbw commented:
@euank you might want https://github.com/kubernetes/kubernetes/pull/25136 too

yifan-gu commented on pkg/kubelet/rkt/container_id.go:
```diff
@@ -41,6 +42,14 @@ func buildContainerID(c *containerID) kubecontainer.ContainerID {
 }
 
+func apiPodToNetPluginContainerID(pod *api.Pod) kubecontainer.ContainerID {
+	return buildContainerID(&containerID{string(pod.UID), pod.Name})
```
Why you use rkt's buildContainerID here? This function generates the runtime container ID, not the API pod UID
I think you want to use the API pod's UID as the network namespace's name, right?

euank replied:
See the implementation of `buildContainerID` above, it always, at the least, injects a `:` we don't want.

I did this via `kubecontainer.ContainerID{ID: string(pod.UID)})` to create a kubecontainer ID that worked correctly with `SetUpPod`

dcbw replied:
I'm not seeing how this is causing errors for me here, since AFAICT the fake container ID made here is only used with SetUpPod/TearDownPod/GetNetNS() and it seems to be round-tripped correctly.  I can't see anything that adds "rkt://" to the kubecontainer.ContainerID unless that got added in the last week since I haven't rebased since then.  It could get added if something did string(<ContainerID>) but that's not something that kubenet or rkt.go appear to do in SetUpPod/TearDownPod...  could have missed it though.

euank commented on pkg/kubelet/rkt/container_id.go:
```diff
@@ -41,6 +42,14 @@ func buildContainerID(c *containerID) kubecontainer.ContainerID {
+}
+
+func runtimePodToNetPluginContainerID(pod *kubecontainer.Pod) kubecontainer.ContainerID {
+	return buildContainerID(&containerID{string(pod.ID), pod.Name})
```
I think this doesn't actually work how we want!

`pod.ID` is of the format `rkt://$uuid`, not `$uuid`.

dcbw replied:
I'm not seeing this, not sure why:

```
I0505 10:17:07.380677   26358 rkt.go:1063] Calling network plugin kubenet to setup pod for hi-openshift_default(6e3c09c0-12d4-11e6-a99d-54ee754bd54c)
W0505 10:17:07.380697   26358 container_id.go:51] $$$$$$$ apiPod container ID &rkt.containerID{uuid:"6e3c09c0-12d4-11e6-a99d-54ee754bd54c", appName:"hi-openshift"}
W0505 10:17:07.380717   26358 container_id.go:52] $$$$$$$ apiPod kube container ID container.ContainerID{Type:"rkt", ID:"6e3c09c0-12d4-11e6-a99d-54ee754bd54c:hi-openshift"}
W0505 10:17:07.380735   26358 kubenet_linux.go:222] ^^^^^ kubenet: container id container.ContainerID{Type:"rkt", ID:"6e3c09c0-12d4-11e6-a99d-54ee754bd54c:hi-openshift"}
W0505 10:17:07.380747   26358 rkt.go:1547] ##### container id container.ContainerID{Type:"rkt", ID:"6e3c09c0-12d4-11e6-a99d-54ee754bd54c:hi-openshift"}
W0505 10:17:07.380760   26358 rkt.go:1548] ##### id &rkt.containerID{uuid:"6e3c09c0-12d4-11e6-a99d-54ee754bd54c", appName:"hi-openshift"}
W0505 10:17:07.380772   26358 rkt.go:1553] ##### string UID "6e3c09c0-12d4-11e6-a99d-54ee754bd54c"
W0505 10:17:07.380782   26358 kubenet_linux.go:227] ^^^^^ kubenet: netns path "/var/run/netns/k8s_6e3c09c0-12d4-11e6-a99d-54ee754bd54c"
I0505 10:17:07.380792   26358 kubenet_linux.go:307] Kubenet: using netns path /var/run/netns/k8s_6e3c09c0-12d4-11e6-a99d-54ee754bd54c
I0505 10:17:07.380797   26358 kubenet_linux.go:308] Kubenet: using podns path default
I0505 10:17:07.380805   26358 kubenet_linux.go:320] Calling cni plugins to add container to network with cni runtime: &{ContainerID:6e3c09c0-12d4-11e6-a99d-54ee754bd54c:hi-openshift NetNS:/var/run/netns/k8s_6e3c09c0-12d4-11e6-a99d-54ee754bd54c IfName:eth0 Args:[]}

```

But I agree that it's very, very confusing with all the IDs flying around.  I'll try to clean it up.

euank commented:
I think during the refactor to pod UID there was some mixup between the 10 `UID` types and 3 `pod` types. As a result, I end up with a pod that never comes up because `SetUpPod` fails with the below.

```
Error syncing pod, skipping: failed to SyncPod: Error adding container to network: Failed to open /var/run/netns/k8-rkt-://f235524a-1231-11e6-a490-42010af00002: open /var/run/netns/k8-rkt-://f235524a-1231-11e6-a490-42010af00002: no such file or directory
```

I'm pretty sure this hits SetUp/TearDown.

dchen1107 commented:
@feiskyer Are you using kubernet today? In Hyper's case, can you return an empty string for GetNetNS for now?

The current CNI model doesn’t work for VMs, and we are aware of there might be issues for hyper. I were talking to @rajatchopra  from redhat who will help us to figure out how network plugins working with VMs in a long run.

feiskyer commented:
@dchen1107 Not yet. As you said, the current CNI model doesn't work for Hypervisors. So `GetNetNS` is ok for now. In the long run, there should be someway to let network plugins also work for Hypervisors. 

> The current CNI model doesn’t work for VMs, and we are aware of there might be issues for hyper. I were talking to @rajatchopra from redhat who will help us to figure out how network plugins working with VMs in a long run.

Thanks. We are figuring how to make CNI work for Hypervisors. @laijs @gnawux

dcbw commented:
@euank @yifan-gu I've cleaned up the ID stuff now.  Still not sure where the rkt:// was getting in, since that never happened for me, but could you try it again?  Thanks!

euank commented:
The UID thing seems to work for me now, but I noticed that I don't have a `lo` interface in my pods with this, while with docker I do see one.

Do you observe this behaviour in your environment? Is CoreOS maybe doing something wonky here?

dcbw commented:
@k8s-bot test this

dcbw commented:
```
[dcbw@localhost kubernetes (kubenet-rkt)]$ sudo ip netns exec k8s_0e2469aa-133f-11e6-8a98-54ee754bd54c ip addr
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
3: eth0@if178: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 82:45:4e:ca:fe:2d brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.1.0.29/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::8045:4eff:feca:fe2d/64 scope link 
       valid_lft forever preferred_lft forever
```

So I do have a loopback, but it's down.  Perhaps you're using ifconfig which doesn't show those.  CNI has a 'loopback' plugin for this, so I'll investigate having kubenet call that before calling the real CNI plugin.

euank commented:
Good call, I indeed was only looking at active interfaces.

The reason it's there in docker, I think, is that the namespace comes from a pause container run with `--net=none` which docker takes to mean "set up a lo interface correctly", not "make a totally barren namespace" as we do (though quite possibly you already know this). The loopback plugin sounds like a good solution. Thanks!

dcbw committed:
kubenet: ensure loopback interface is up and consolidate CNI setup

dcbw commented:
@euank how about now?

dcbw commented:
@k8s-bot test this issue: #25153

euank commented:
This LGTM. Thanks a load @dcbw!

There might still be some rough edges, but I'd like to get this merged and improve from there if no one has major concerns.

dcbw committed:
kubenet: fix log message text

dcbw committed:
kubenet: implement fake execer for testcases

When the IP isn't in the internal map, GetPodNetworkStatus() needs
to call the execer for the 'nsenter' program.  That  means the execer
needs to be !nil, which it wasn't before.

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -937,15 +985,21 @@ func (r *Runtime) preparePod(pod *api.Pod, pullSecrets []api.Secret) (string, *k
 		// This enables graceful stop.
 		newUnitOption("Service", "KillMode", "mixed"),
+		// Track pod info for garbage collection
+		newUnitOption(unitKubernetesSection, unitPodUID, string(pod.UID)),
```
@dcbw I think we already store these information in the annotations in the rkt pod manifest. I don't like we store information in multiple places. 
But this is good for now, I will need to do a rebase and clean up for #24647 after this gets in. cc @euank

euank replied:
I don't have a problem with storing information in multiple locations in this case since it's useful for correlating distinct resources.

dcbw replied:
I chose to store it the systemd service file too since that's already found for Garbage Collect.  Also, the manifest file is removed from the tempdir after preparePod(), so it's not guaranteed to be around during GC, but the systemd unit will be.  Right?  Or is there another way to get at the pod manifest given the systemd unit file?

yifan-gu replied:
@dcbw I mean we can read the podmanifest in the response of the ListPods() or InspectPod() from rkt api service.
OOB discussion with @euank, I agree that this can be more user-friendly for debugging. But for developers who will work on the code, I would say a single source of truth would be easy to manage.
However, I won't block the PR on this.

dcbw replied:
@yifan-gu agreed, that sounds like a good future direction for the GC code.

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -864,20 +885,47 @@ func (r *Runtime) generateRunCommand(pod *api.Pod, uuid string) (string, error)
+func (r *Runtime) cleanupPodNetwork(pod *api.Pod) error {
+	podDetail := fmt.Sprintf("%s_%s(%s)", pod.Name, pod.Namespace, pod.UID)
+
+	glog.V(3).Infof("Calling network plugin %s to tear down pod for %s", r.networkPlugin.Name(), podDetail)
```
s/podDetail/format.Pod(pod)

dcbw replied:
@yifan-gu Fixed in latest push.

yifan-gu commented:
@euank Have we e2e test this PR?

euank commented:
@yifan-gu unfortunately not yet.. the dream of further automation is still unrealized.

dcbw commented:
@euank @yifan-gu @dchen1107 are we ready to go on this one?  anything else I should do?

yifan-gu commented:
@dcbw rkt changes LGTM
cc @bprashanth for a final look?

yifan-gu commented:
Ping @bprashanth @thockin

yujuhong commented:
/cc @freehan who's working on the kubenet stuff.

yifan-gu commented:
Sorry, my pr gets in first, so we need a rebase here :)

yifan-gu commented:
Ping @freean @bprashanth @thockin @kubernetes/sig-network , PTAL on the kubenet change, thank you :)

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1547,8 +1715,22 @@ func (r *Runtime) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy) error
+				serviceFile := serviceFilePath(f.Name())
+				id, name, namespace, err := podDetailsFromServiceFile(serviceFile)
+				if err == nil {
+					r.cleanupPodNetwork(&api.Pod{
```
@dcbw So this only cleans up the pod network for orphanage service. We recently added the removePod() below to remove the actual pod, so you probably want to cleanupPodNetwork there as well :)

freehan replied:
Everytime there is an error, we invoke cleanupPodNetwork. Is it possible that we defer a validation function? In this function, check if the current operation succeeded, and then invoke cleanupPodNetwork if needed.

freehan commented on pkg/kubelet/network/hairpin/hairpin.go:
```diff
@@ -40,21 +40,26 @@ var (
 }
 
-func findPairInterfaceOfContainerInterface(e exec.Interface, containerPid int, containerInterfaceName string) (string, error) {
+func findPairInterfaceOfContainerInterface(e exec.Interface, containerPid int, netnsPath, containerInterfaceName string) (string, error) {
```
Mabe just separate this function into findPairInterfaceByPid and findPairInterfaceByNetnsPath

freehan replied:
Or add a comment about the expected input 

Also apply to setUpContainerInternal

freehan commented:
some nits. overall LGTM

dcbw commented:
@freehan how about the updated hairpin commit?

dcbw commented:
@yifan-gu I also addressed the removePod() thing; PTAL thanks!

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1581,18 +1755,42 @@ func (r *Runtime) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy) error
+		return fmt.Errorf("rkt: Failed to read pod details from systemd service file %v: %v", serviceFilePath, err)
+	}
+	
+	r.cleanupPodNetwork(&api.Pod{
```
why not we return the error here?

yifan-gu replied:
btw, we cleanup pod network both in KillPod() and GC(), so maybe the error can be ignored?

dcbw replied:
The network may not be configured at this time, so a fatal error here should not fail cleaning up the exited pod.  But I'll change the error return to be ignored.

yifan-gu commented on pkg/kubelet/container/runtime.go:
```diff
@@ -97,6 +97,10 @@ type Runtime interface {
+	// Returns the filesystem path of the pod's network namespace; if the
+	// runtime does not handle namespace creation itself, or cannot return
+	// the network namespace path, it should return an error.
+	GetNetNS(containerID ContainerID) (string, error)
```
nits: Can we add a TODO here to say we want to revisit this and change ContainerID to pod UID in the future?

yujuhong replied:
TODO sounds good. We'll probably include this via the pod sandbox interface in the future, so container id is not needed.

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -972,15 +1017,21 @@ func (r *Runtime) preparePod(pod *api.Pod, pullSecrets []api.Secret) (string, *k
 	glog.V(4).Infof("rkt: Creating service file %q for pod %q", serviceName, format.Pod(pod))
 	serviceFile, err := r.os.Create(serviceFilePath(serviceName))
 	if err != nil {
+		r.cleanupPodNetwork(pod)
```
I think we can delete this, as it's taken care of in the parent function RunPod()

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -972,15 +1017,21 @@ func (r *Runtime) preparePod(pod *api.Pod, pullSecrets []api.Secret) (string, *k
 		return "", nil, err
 	}
 	if _, err := io.Copy(serviceFile, unit.Serialize(units)); err != nil {
+		r.cleanupPodNetwork(pod)
```
ditto

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1357,6 +1457,30 @@ func (r *Runtime) KillPod(pod *api.Pod, runningPod kubecontainer.Pod, gracePerio
+			return err
+		}
+	}
+
```
Can we move the cleanup code snippets after `if res != "done"` ?
Also seems the `podUID`, `podName`, `podNamespace` are the only required fields, so how about simplify the code as:
```go
netCleanupPod = &api.Pod{
    ObjectMeta: api.ObjectMeta{
        UID:       runningPod.ID,
        Name:      runningPod.Name,
        Namespace: runningPod.Namespace,
    },
}

// TODO: Check if the running pod is host network pod.
if pod == nil || !kubecontainer.IsHostNetworkPod(pod) {
    if err := cleanupPodNetwork(netCleanupPod); err != nil {
        ...
    }
}

yifan-gu commented:
Sorry, added a few more nits :stuck_out_tongue:

dcbw commented:
@yifan-gu PTAL, thanks!

yifan-gu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -899,20 +919,45 @@ func (r *Runtime) generateRunCommand(pod *api.Pod, uuid string) (string, error)
+	}
+
+	if _, err := r.execer.Command("ip", "netns", "del", makePodNetnsName(pod.UID)).Output(); err != nil {
+		return fmt.Errorf("rkt: Failed to remove  network namespace for pod %s: %v", format.Pod(pod), err)
```
nit: one extra space in `remove  network`.

yifan-gu commented:
One small nits, LGTM.

dcbw commented:
> One small nit, LGTM.

@yifan-gu fixed

yifan-gu commented:
LGTM. Thank you very much @dcbw !

euank commented:
It's not clear to me if those three failures are caused by this or flakes, but since one of them could relate to the ip in podStatus (`kubelet managed /etc/hosts file`) and one is networking (healthz), it's a little scary.

bprashanth commented:
Hmm all 3 instances looks like the test failed trying to contact apiserver:
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25062/kubernetes-pull-build-test-e2e-gce/40667/
```
  Expected error:
      <*url.Error | 0xc820509dd0>: {
          Op: "Get",
          URL: "https://104.154.95.7/api/v1/watch/namespaces/e2e-tests-e2e-kubelet-etc-hosts-4a3o9/serviceaccounts?fieldSelector=metadata.name%3Ddefault",
          Err: {
              Op: "dial",
              Net: "tcp",
              Source: nil,
              Addr: {
                  IP: "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xffh\x9a_\a",
                  Port: 443,
                  Zone: "",
              },
              Err: {},
          },
      }


/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:508
  should *not* be restarted with a /healthz http liveness probe [Conformance] [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:1164

  creating pod liveness-http
  Expected error:
      <*url.Error | 0xc82038cf00>: {
          Op: "Post",
          URL: "https://104.154.95.7/api/v1/namespaces/e2e-tests-pods-6nwro/pods",
          Err: {
              Op: "dial",
              Net: "tcp",
              Source: nil,
              Addr: {
                  IP: "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xffh\x9a_\a",
                  Port: 443,
                  Zone: "",
              },
              Err: {},
          },
      }
      Post https://104.154.95.7/api/v1/namespaces/e2e-tests-pods-6nwro/pods: dial tcp 104.154.95.7:443: i/o timeout
  not to have occurred

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:508
  should update labels on modification [Conformance] [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/downwardapi_volume.go:96

  Expected error:
      <*url.Error | 0xc8205d0690>: {
          Op: "Post",
          URL: "https://104.154.95.7/api/v1/namespaces/e2e-tests-downward-api-7q9y5/pods",
          Err: {
              Op: "dial",
              Net: "tcp",
              Source: nil,
              Addr: {
                  IP: "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xffh\x9a_\a",
                  Port: 443,
                  Zone: "",
              },
              Err: {},
          },
      }
      Post https://104.154.95.7/api/v1/namespaces/e2e-tests-downward-api-7q9y5/pods: dial tcp 104.154.95.7:443: i/o timeout
  not to have occurred
```

yifan-gu commented:
As a notes. The `defaultNetworkName` can be dropped now after this PR. And we need the update the `GetPodStatus()` to return the correct pod IP after this PR is merged. cc @euank

euank commented:
For reference, two of the commits [here](https://github.com/euank/kubernetes/commits/rkt-pod-ip) are probably what's needed, I'm testing to make sure that didn't break anything. If possible, I want to just get this PR's changes in so we can continue iterating, but if it's better for the `networkPlugin.GetPodNetworkStatus` related bits to be here (in terms of speed and/or correctness), that's fine too.

freehan commented:
LGTM. pending on tests.

euank commented:
An apiserver failure seems unrelated to this, but I'm not sure enough it's a flake to file a flake issue yet. I'll file an issue if the below retest passes, proving it a flake.

... And the above is a catch 22 for retesting :smile: 

@k8s-bot test this issue: #IGNORE 

([failure link](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25062/kubernetes-pull-build-test-e2e-gce/40667/) in case it needs to be preserved in the case of it being a flake)

dcbw commented:
@euank the pass-pod-IP commit looks OK to me, my mistake to have missed that originally...
